### DISCUSSION TITLE
[@types/node] Wrong type for buffer argument in Node.js crypto

### DISCUSSION THREAD

**Discussion by Ceres6** [DefinitelyTyped/DefinitelyTyped#70790] [Issues with a @types package]
Currently the method `crypto.publicEncrypt` is defined as 

```function publicEncrypt(key: RsaPublicKey | RsaPrivateKey | KeyLike, buffer: NodeJS.ArrayBufferView): Buffer;```

[here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/crypto.d.ts#L2142C5-L2142C113) but according to [Node.js documentation](https://nodejs.org/docs/latest/api/crypto.html#cryptopublicencryptkey-buffer) it also accepts `string` and `Buffer`.

 I'm using the `Buffer` input in my project and it works okay. It wasn't complaining until I upgraded TS from 4.9 to 5.6 so probably there have been some changes there. In any case I think we should stick to Node.js docs. I'm happy to push a PR for this but wanted to check if I might be missing something. There are other methods in crypto with the same issue, I could include those in a PR as well.

Thanks in advance.

[a comment by an automated account is not shown]

**Comment by Renegade334** [comment]
I suspect this will end up bisecting to #70694. Can you confirm exactly what versions of TypeScript and @types/node you're using?

`Buffer` should be included in `NodeJS.ArrayBufferView` (`Buffer` extends `Uint8Array` extends `NodeJS.TypedArray` extends `NodeJS.ArrayBufferView`) so this is an assignability issue rather than a question of the type being missing.

(FWIW, `string` _is_ missing from the `buffer` argument to the public/private encrypt/decrypt functions, as it's a valid argument to [`getArrayBufferOrView()`](https://github.com/nodejs/node/blob/v22.9.0/lib/internal/crypto/util.js#L138-L160).)

**Reply by Ceres6** [reply]
Yes it seems to be an assignability issue due to the returned types for `entries()`. 

I was using `typescript: 5.6.2` and `@types/node: 18.15.5`, updating to `@types/node: 18.19.54` seems to solve the issue, but it probably breaks other stuff as I'm not using node `18.19`, not sure if the fix can be backported to `v18.15`

About the missing string type, should I make a PR adding it?

**Reply by Renegade334** [reply]
So this is actually related to microsoft/TypeScript#59996, in that (intentionally or unintentionally) you're loading lib.esnext.iterator in TS 5.6+, which is incompatible with certain interfaces in @types/node that have methods that overlap with iterator-returning builtins.

If you're deliberately supporting an older LTS version of Node, then you almost certainly don't want to be compiling against `target: esnext`. Change your `target` or `lib` tsconfig entry to avoid loading the esnext iterator declarations, and your error in v18.15.x should disappear.

> About the missing string type, should I make a PR adding it?

Sure 👍

**Reply by Ceres6** [reply]
> If you're deliberately supporting an older LTS version of Node, then you almost certainly don't want to be compiling against `target: esnext`. Change your `target` or `lib` tsconfig entry to avoid loading the esnext iterator declarations, and your error in v18.15.x should disappear.

I think I might be missing something, because I have 
```json
{
 "compilerOptions": {
    "lib": ["es2022"],
    "target": "es2022"
  }
}
```

EDIT: My build worked okay, ~it was just my vscode `implicit project config: Module` that was loading `esnext`~. 

EDIT2: it seems the problem is with my vscode detection of errors, but Idk what is it

Thanks for your help, I'll try to get the string PR up today

**Reply by Ceres6** [reply]
It seems my problem is that `@types/jest` is creeping into my editor type checking, not sure if I should open a different discussion for that
